### ISSUE TITLE
IWYU incorrect forward declare

### ISSUE THREAD

**Issue by jcageman** [include-what-you-use/include-what-you-use#425]
Repro:

DataElement.h

```
class CDataElement
{
};

```
main.cpp

```
#include <tchar.h>
#include "DataElement.h"


template < class T >
class CCreator
{
public:
	static void* Create()
	{
		return new T;
	}
};

typedef void* (*TConstructor)(void);

void RegisterType( TConstructor /*i_pfnConstructor*/ )
{
}

int _tmain(int argc, _TCHAR* argv[])
{
	RegisterType(CCreator<CDataElement>::Create);
	return 0;
}
```

IWYU suggestions:

```
D:/src/IWYUReproProject/IWYUReproProject/IWYUReproProject.cpp should add these lines:
class CDataElement;

D:/src/IWYUReproProject/IWYUReproProject/IWYUReproProject.cpp should remove these lines:
- #include "DataElement.h"  //lines 5-5

The full include-list for D:/src/IWYUReproProject/IWYUReproProject/IWYUReproProject.cpp:
#include <tchar.h>
class CDataElement;
---

```

applying the suggestions results in the following compile error in visual studio 2012
` 'CDataElement' : no appropriate default constructor available`

If you need more logging let me know.

If i change:
`RegisterType(CCreator<CDataElement>::Create);`
to
`RegisterType(&CCreator<CDataElement>::Create);`

 iwyu will make the correct decision by the way. E.g.  the & is crucial here. I have to admit with & would make more sense, but its legal C++ i think.

I tried debugging it myself, but i have to admit i do not understand the AST parser well enough to fix it.

**Comment by jcageman** [comment]
An update on this issue:

I managed to actually understand a bit. Like i mentioned previously the difference between & and without is crucial here so i just made an ast-dump and diffed the dumps.

Without &
```
 | | | |-ExprWithCleanups <line:116:4, col:87> 'void'
  | | | | `-CXXMemberCallExpr <col:4, col:87> 'void'
  | | | |   |-MemberExpr <col:4, col:7> '<bound member function type>' .RegisterCompositeConstructor   | | | |   | `-DeclRefExpr <col:4> 'class CParameterHolder' lvalue Var 'PH' 'class CParameterHolder'
  | | | |   |-CXXConstructExpr <col:37, col:46> 'class CParameter::TToken':'class CStringToken' 'void (class CStringToken &&) __attribute__((thiscall)) noexcept' elidable
  | | | |   | `-MaterializeTemporaryExpr <col:37, col:46> 'class CParameter::TToken':'class CStringToken' xvalue
  | | | |   |   `-CXXMemberCallExpr <col:37, col:46> 'class CParameter::TToken':'class CStringToken'
  | | | |   |     `-MemberExpr <col:37> '<bound member function type>' ->GetToken   | | | |   |       `-ImplicitCastExpr <col:37> 'const class CDataElementsKeeper::ABCDataElement *' <UncheckedDerivedToBase (ABCDataElement)>
  | | | |   |         `-CXXThisExpr <col:37> 'class DataElementRegistration::CConfiguration *' this
  | | | |   `-ImplicitCastExpr <col:56, col:80> 'class CParameter *(*)(void)' <FunctionToPointerDecay>
  | | | |     `-DeclRefExpr <col:56, col:80> 'class CParameter *(void)' lvalue CXXMethod 'Create' 'class CParameter *(void)'

```

With &
```
 | | |-ExprWithCleanups <line:116:4, col:88> 'void'
  | | | | `-CXXMemberCallExpr <col:4, col:88> 'void'
  | | | |   |-MemberExpr <col:4, col:7> '<bound member function type>' .RegisterCompositeConstructor   | | | |   | `-DeclRefExpr <col:4> 'class CParameterHolder' lvalue Var 'PH' 'class CParameterHolder'
  | | | |   |-CXXConstructExpr <col:37, col:46> 'class CParameter::TToken':'class CStringToken' 'void (class CStringToken &&) __attribute__((thiscall)) noexcept' elidable
  | | | |   | `-MaterializeTemporaryExpr <col:37, col:46> 'class CParameter::TToken':'class CStringToken' xvalue
  | | | |   |   `-CXXMemberCallExpr <col:37, col:46> 'class CParameter::TToken':'class CStringToken'
  | | | |   |     `-MemberExpr <col:37> '<bound member function type>' ->GetToken   | | | |   |       `-ImplicitCastExpr <col:37> 'const class CDataElementsKeeper::ABCDataElement *' <UncheckedDerivedToBase (ABCDataElement)>
  | | | |   |         `-CXXThisExpr <col:37> 'class DataElementRegistration::CConfiguration *' this
  | | | |   `-UnaryOperator <col:56, col:81> 'class CParameter *(*)(void)' prefix '&'
  | | | |     `-DeclRefExpr <col:57, col:81> 'class CParameter *(void)' lvalue CXXMethod 'Create' 'class CParameter *(void)'
```
Here the difference is the fact that the & is a UnaryOperator, while without it its a ImplicitCastExpr

I took a look in the code for TraverseDeclRefExpr in iwyu.cc

I changed:

```
    if (current_ast_node()->template ParentIsA<CallExpr>() || current_ast_node()->template ParentIsA<ImplicitCastExpr>() && current_ast_node()->template AncestorIsA<CallExpr>(2) ) {
      return true;
    }
```

to
```
    if (current_ast_node()->template ParentIsA<CallExpr>() ) {
      return true;
    }

	if (current_ast_node()->template ParentIsA<ImplicitCastExpr>() && current_ast_node()->template AncestorIsA<CallExpr>(2) )
	{
		const ImplicitCastExpr* imp_cast_expr = current_ast_node()->template GetParentAs<ImplicitCastExpr>();
		if (imp_cast_expr && imp_cast_expr->getCastKind() != clang::CK_FunctionToPointerDecay)
		{
			return true;
		}
	}
```

That does the job :) I have no clue if it actually correct, but it feels like at least close to the correct solution.

**Comment by kimgr** [comment]
Cool, thanks! I meant to respond to this, but I lost it.

Since the AST appears to have changed, or is slightly different than expected in this contetxt, I think the change is in the right place. We should get a test case in place for this based on your example.

**Comment by kimgr** [comment]
It looks like this patch will also fix the problem:

```diff
diff --git a/iwyu.cc b/iwyu.cc
index 1c4297a..2bcbeeb 100644
--- a/iwyu.cc
+++ b/iwyu.cc
@@ -898,21 +898,13 @@ class BaseAstVisitor : public RecursiveASTVisitor<Derived> {
         const_cast<CXXDestructorDecl*>(dtor), parent_type, expr);
   }

-  // This is to catch assigning template functions to function pointers.
+  // This is to catch binding of function pointers to templates.
   // For instance, 'MyFunctionPtr p = &TplFn<MyClass*>;': we need to
   // expand TplFn to see if it needs full type info for MyClass.
   bool TraverseDeclRefExpr(clang::DeclRefExpr* expr) {
     if (!Base::TraverseDeclRefExpr(expr))  return false;
     if (CanIgnoreCurrentASTNode())  return true;

-    // If it's a normal function call, that was already handled by a
-    // CallExpr somewhere.  We want only assignments.
-    if (current_ast_node()->template ParentIsA<CallExpr>() ||
-        (current_ast_node()->template ParentIsA<ImplicitCastExpr>() &&
-         current_ast_node()->template AncestorIsA<CallExpr>(2))) {
-      return true;
-    }
-
     if (FunctionDecl* fn_decl = DynCastFrom(expr->getDecl())) {
       // If fn_decl has a class-name before it -- 'MyClass::method' --
       // it's a method pointer.

```

That is, we can remove the looking-up-the-tree -- there's no test coverage for it anyway.

And the comment was lying, CallExpr only handles references in varargs for function calls, so it misses the fact that a function-pointer argument requires the full type, including template args.

It's getting late here, but tomorrow I'll try to reduce your example to a nice IWYU test case so we can get this under test.

Thanks for the help!

**Comment by kimgr** [comment]
Fixed in PR #429.
